dpx/src/specials/util.rs

```diff
@@ -263,30 +268,28 @@ unsafe extern "C" fn spc_read_color_color(
  */
 unsafe extern "C" fn spc_read_color_pdf(
     mut spe: *mut spc_env,
-    colorspec: &mut pdf_color,
     mut ap: *mut spc_arg,
-) -> i32 {
+) -> Result<PdfColor, ()> {
     let mut cv: [f64; 4] = [0.; 4]; /* at most four */
     let mut nc: i32 = 0;
-    let mut isarry: i32 = 0i32;
-    let mut error: i32 = 0i32;
+    let mut isarry: bool = false;
     let mut q: *mut i8 = 0 as *mut i8;
     skip_blank(&mut (*ap).curptr, (*ap).endptr);
     if *(*ap).curptr.offset(0) as i32 == '[' as i32 {
         (*ap).curptr = (*ap).curptr.offset(1);
         skip_blank(&mut (*ap).curptr, (*ap).endptr);
-        isarry = 1i32
+        isarry = true
     }
     nc = spc_util_read_numbers(cv.as_mut_ptr(), 4i32, ap);
-    match nc {
+    let mut result = match nc {
         1 => {
-            pdf_color_graycolor(colorspec, cv[0]);
+            PdfColor::from_gray(cv[0]).map_err(|err| err.warn())
```

Comment:
Stop. The problem is still here.
If new value is incorrect, it should warn and keep old value (not replace it with incorrect).

Reply:
or you can:
```rust
PdfColor::from_gray(cv[0]).unwrap_or_else(|err| { err.warn(), colorspec })
```

Reply:
Although I may be wrong.

Reply:
Here the supplied color was always an invalid color created by the `new` method. I have replaced it by this function which will return `Err(())`. Since the code only used the color value if there was no error the semantics are the same, with the added benefit that it is now impossible to use an invalid color.